WebArchive content should not terminate web process when accessing cookies

achristensen07 opened the pull request:
#### 9c2ad9a13f9314412d153d7ab7b1444ac74290c7
<pre>
WebArchive content should not terminate web process when accessing cookies
<a href="https://bugs.webkit.org/show_bug.cgi?id=250537">https://bugs.webkit.org/show_bug.cgi?id=250537</a>
rdar://103003585

Reviewed by Chris Dumez.

There are 3 different ways I&apos;m aware of to load WebArchives.  All of them needed fixing.
When we load a web archive, it is only parsed in the web process, which is the only place
we can find out what domain it was originally fetched from, which is contained inside the
web archive data, which only has a way to parse it in its entirety.  Rather than parse
the whole thing in the UI process or buffer it in the network process memory then parse it,
I just disable cookie domain checks when a web archive has been loaded, which is rare.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::maybeCreateArchive):
* Source/WebCore/loader/FrameLoaderClient.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::allowCookieAccessForWebArchive):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::initializeNetworkProcess):
(WebKit::NetworkProcess::addAllowedFirstPartyForCookies):
(WebKit::NetworkProcess::webProcessWillLoadWebArchive):
(WebKit::NetworkProcess::webProcessDidLoadWebArchive):
(WebKit::NetworkProcess::allowsFirstPartyForCookies):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::didReceiveMainResourceResponse):
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::ensureSWServer):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.cpp:
(WebKit::WebSharedWorkerServer::createContextConnection):
* Source/WebKit/Shared/LoadedWebArchive.h: Added.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadAlternateHTML):
(WebKit::WebPageProxy::loadWebArchiveData):
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::allowCookieAccessForWebArchive):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: <a href="https://commits.webkit.org/258901@main">https://commits.webkit.org/258901@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be3730722f44e86ae9da44d442609e7876edac08

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103244 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12369 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36231 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112490 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172688 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107197 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13395 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3273 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95474 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/110737 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109018 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10292 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93427 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37929 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92110 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24972 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79655 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5769 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26378 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5943 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2887 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11929 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45881 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7691 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3259 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
```diff
@@ -1494,6 +1494,11 @@ void WebFrameLoaderClient::prepareForDataSourceReplacement()
 
+void WebFrameLoaderClient::allowCookieAccessForWebArchive(const URL& firstParty)
+{
+    WebProcess::singleton().ensureNetworkProcessConnection().connection().send(Messages::NetworkConnectionToWebProcess::AllowCookieAccessForWebArchive(firstParty), 0);
```
The WebProcess is not a trusted process. Why is it OK to have an IPC from the WebProcess to the network process to relax cookie access restrictions?

achristensen07 replied:
This is described in the commit message.  It's only allowed if an external process has verified that a web archive has gone into the web process.  All other possible solutions involve getting the entire web archive in memory or redoing the parser, and parsing the archive outside of the web process to get the "first party" from the web archive contents.  This was the most efficient solution and it doesn't introduce new parsing into other processes.

achristensen07 replied:
This line makes it safe:
NETWORK_PROCESS_MESSAGE_CHECK(m_networkProcess->webProcessDidLoadWebArchive(m_webProcessIdentifier));

cdumez replied:
Well, "safe". It checks that the WebProcess actually loaded a WebArchive but totally blindly trusts the domain sent by the insecure WebProcess it seems.

cdumez reviewed (approved):
LGTM

achristensen07 committed:
WebArchive content should not terminate web process when accessing cookies
https://bugs.webkit.org/show_bug.cgi?id=250537
rdar://103003585

Reviewed by Chris Dumez.

There are 3 different ways I'm aware of to load WebArchives.  All of them needed fixing.
When we load a web archive, it is only parsed in the web process, which is the only place
we can find out what domain it was originally fetched from, which is contained inside the
web archive data, which only has a way to parse it in its entirety.  Rather than parse
the whole thing in the UI process or buffer it in the network process memory then parse it,
I just disable cookie domain checks when a web archive has been loaded, which is rare.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::maybeCreateArchive):
* Source/WebCore/loader/FrameLoaderClient.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::allowCookieAccessForWebArchive):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::initializeNetworkProcess):
(WebKit::NetworkProcess::addAllowedFirstPartyForCookies):
(WebKit::NetworkProcess::webProcessWillLoadWebArchive):
(WebKit::NetworkProcess::webProcessDidLoadWebArchive):
(WebKit::NetworkProcess::allowsFirstPartyForCookies):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::didReceiveMainResourceResponse):
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::ensureSWServer):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.cpp:
(WebKit::WebSharedWorkerServer::createContextConnection):
* Source/WebKit/Shared/LoadedWebArchive.h: Added.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadAlternateHTML):
(WebKit::WebPageProxy::loadWebArchiveData):
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::allowCookieAccessForWebArchive):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: https://commits.webkit.org/258901@main
